File: DotNetMessaging/SpringBootMessaging/src/main/java/com/messaging/api/repositories/ChatRepository.java
package com.messaging.api.repositories;

import com.messaging.api.models.Chat;
import org.springframework.data.mongodb.repository.MongoRepository;
import org.springframework.data.mongodb.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface ChatRepository extends MongoRepository<Chat, String> {
    @Query("{ $or: [ { user1Id: ?0, user2Id: ?1 }, { user1Id: ?1, user2Id: ?0 } ] }")
    Optional<Chat> findByUserIds(String user1Id, String user2Id);
    
    @Query("{ $or: [ { user1Id: ?0 }, { user2Id: ?0 } ] }")
    List<Chat> findByUserId(String userId);
}

File: DotNetMessaging/backend/DotNetMessaging.API/Repositories/ChatRepository.cs
using MongoDB.Driver;
using DotNetMessaging.API.Data;
using DotNetMessaging.API.Models;

namespace DotNetMessaging.API.Repositories;

public class ChatRepository : Repository<Chat>, IChatRepository
{
    public ChatRepository(MongoDbContext context) : base(context, "chats")
    {
        // Create compound index for user pairs
        var indexKeys = Builders<Chat>.IndexKeys
            .Ascending(c => c.User1Id)
            .Ascending(c => c.User2Id);
        var indexOptions = new CreateIndexOptions { Unique = true };
        _collection.Indexes.CreateOne(new CreateIndexModel<Chat>(indexKeys, indexOptions));
    }

    public async Task<Chat?> GetChatBetweenUsersAsync(string user1Id, string user2Id)
    {
        var filter = Builders<Chat>.Filter.Or(
            Builders<Chat>.Filter.And(
                Builders<Chat>.Filter.Eq(c => c.User1Id, user1Id),
                Builders<Chat>.Filter.Eq(c => c.User2Id, user2Id)
            ),
            Builders<Chat>.Filter.And(
                Builders<Chat>.Filter.Eq(c => c.User1Id, user2Id),
                Builders<Chat>.Filter.Eq(c => c.User2Id, user1Id)
            )
        );
        
        return await _collection.Find(filter).FirstOrDefaultAsync();
    }

    public async Task<IEnumerable<Chat>> GetUserChatsAsync(string userId)
    {
        var filter = Builders<Chat>.Filter.Or(
            Builders<Chat>.Filter.Eq(c => c.User1Id, userId),
            Builders<Chat>.Filter.Eq(c => c.User2Id, userId)
        );
        
        var sort = Builders<Chat>.Sort.Descending(c => c.LastMessageAt).Descending(c => c.CreatedAt);
        
        return await _collection.Find(filter).Sort(sort).ToListAsync();
    }
}

